engine/src/main/java/org/pentaho/di/job/entries/mail/JobEntryMail.java

```diff
@@ -0,0 +1,1366 @@
+            endRow );
+          messageText.append( "------------------------" ).append( endRow );
+
+          addBacktracking( jobTracker, messageText );
+          if ( isUseHTML() ) {
+            messageText.replace( 0, messageText.length(), messageText.toString().replace( Const.CR, endRow ) );
+          }
+        }
+      }
+
+      MimeMultipart parts = new MimeMultipart();
+      MimeBodyPart part1 = new MimeBodyPart(); // put the text in the
+      // Attached files counter
+      int nrattachedFiles = 0;
+
+      // 1st part
+
+      if ( useHTML ) {
+        if ( !Utils.isEmpty( getEncoding() ) ) {
+          part1.setContent( messageText.toString(), "text/html; " + "charset=" + getEncoding() );
+        } else {
+          part1.setContent( messageText.toString(), "text/html; " + "charset=ISO-8859-1" );
+        }
+      } else {
+        part1.setText( messageText.toString() );
+      }
+
+      parts.addBodyPart( part1 );
+
+      if ( includingFiles && result != null ) {
+        List<ResultFile> resultFiles = result.getResultFilesList();
+        if ( resultFiles != null && !resultFiles.isEmpty() ) {
+          if ( !zipFiles ) {
+            // Add all files to the message...
+            //
+            for ( ResultFile resultFile : resultFiles ) {
+              FileObject file = resultFile.getFile();
+              if ( file != null && file.exists() ) {
+                boolean found = false;
+                for ( int i = 0; i < fileType.length; i++ ) {
+                  if ( fileType[i] == resultFile.getType() ) {
+                    found = true;
+                  }
+                }
+                if ( found ) {
+                  // create a data source
+                  MimeBodyPart files = new MimeBodyPart();
+                  URLDataSource fds = new URLDataSource( file.getURL() );
+
+                  // get a data Handler to manipulate this file type;
+                  files.setDataHandler( new DataHandler( fds ) );
+                  // include the file in the data source
+                  files.setFileName( file.getName().getBaseName() );
+
+                  // insist on base64 to preserve line endings
+                  files.addHeader( "Content-Transfer-Encoding", "base64" );
+
+                  // add the part with the file in the BodyPart();
+                  parts.addBodyPart( files );
+                  nrattachedFiles++;
+                  logBasic( "Added file '" + fds.getName() + "' to the mail message." );
+                }
+              }
+            }
+          } else {
+            // create a single ZIP archive of all files
+            masterZipfile =
+              new File( System.getProperty( "java.io.tmpdir" )
+                + Const.FILE_SEPARATOR + environmentSubstitute( zipFilename ) );
+            ZipOutputStream zipOutputStream = null;
+            try {
+              zipOutputStream = new ZipOutputStream( new FileOutputStream( masterZipfile ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *zipOutputStream*. The resource is referenced at lines 1040, 1052, 1059, 1074, 1075, 1077. The resource is closed at lines 1064, 1078. There are other execution paths that do not contain closure statements, for example, when ZipOutputStream.finish throws an exception. Ensure *zipOutputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).